packages/recommend-react/src/RecommendContext.tsx

```diff
@@ -0,0 +1,207 @@
+import { RecommendationsQuery, RecommendClient } from '@algolia/recommend';
+import { BatchKeyPair } from '@algolia/recommend-core';
+import {
+  BatchRecommendations,
+  getBatchRecommendations,
+} from '@algolia/recommend-core/src';
+import { dequal } from 'dequal';
+import React from 'react';
+
+import { isPresent } from './utils/isPresent';
+
+type GetParametersResult = {
+  queries: RecommendationsQuery[];
+  keyPair: BatchKeyPair;
+};
+
+type RecommendWidget<TObject> = {
+  getParameters: () => GetParametersResult;
+  onResult: (value: BatchRecommendations<TObject>) => void;
+  onRequest: () => void;
+  key: string;
+  param: GetParametersResult;
+};
+
+type RecommendContextType<TObject> = {
+  register: (
+    widget: Omit<RecommendWidget<TObject>, 'param'>
+  ) => (key: string) => void;
+  hasProvider: boolean;
+  recommendClient: RecommendClient;
+};
+
+type RecommendProviderProps = {
+  children: React.ReactNode;
+  recommendClient: RecommendClient;
+};
+
+type StateType<TObject> = {
+  isDirty: number | null;
+  cache: Record<string, BatchRecommendations<TObject>>;
+  widgets: Array<RecommendWidget<TObject>>;
+  recommendClient: RecommendClient | null;
+};
+
+type Action<TObject> =
+  | { type: 'register'; widget: Omit<RecommendWidget<TObject>, 'param'> }
+  | { type: 'unregister'; key: string }
+  | { type: 'request_success' };
+
+export const RecommendContext = React.createContext<
```

Comment:
For better debugging experience in the DevTools, we can add a `displayName` to the context in development. See [how it's done in Predict](https://github.com/algolia/predict/blob/c5040758c6807e6dd5d5b10d1e7d7f463514b353/packages/predict-react/src/lib/PredictContext.ts#L65-L67).